src/Lib.P2P/PubSub/PubSubService.cs

```diff
@@ -0,0 +1,278 @@
+        /// <param name="data">
+        ///   The payload of message.
+        /// </param>
+        /// <returns>
+        ///   A unique published message.
+        /// </returns>
+        /// <remarks>
+        ///   The <see cref="PublishedMessage.SequenceNumber"/> is a monitonically 
+        ///   increasing unsigned long.
+        /// </remarks>
+        public PublishedMessage CreateMessage(string topic, byte[] data)
+        {
+            var next = Interlocked.Increment(ref _nextSequenceNumber);
+            var seqno = BitConverter.GetBytes(next);
+            if (BitConverter.IsLittleEndian)
+            {
+                seqno = seqno.Reverse().ToArray();
+            }
+            
+            return new PublishedMessage
+            {
+                Topics = new[] {topic},
+                Sender = LocalPeer,
+                SequenceNumber = seqno,
+                DataBytes = data
+            };
+        }
+
+        /// <inheritdoc />
+        public Task<IEnumerable<string>> SubscribedTopicsAsync(CancellationToken cancel = default)
+        {
+            var topics = _topicHandlers.Values
+               .Select(t => t.Topic)
+               .Distinct();
+            return Task.FromResult(topics);
+        }
+
+        /// <inheritdoc />
+        public Task<IEnumerable<Peer>> PeersAsync(string topic = null,
+            CancellationToken cancel = default)
+        {
+            var peers = Routers
+               .SelectMany(r => r.InterestedPeers(topic))
+               .Distinct();
+            return Task.FromResult(peers);
+        }
+
+        /// <inheritdoc />
+        public Task PublishAsync(string topic, string message, CancellationToken cancel = default)
+        {
+            return PublishAsync(topic, Encoding.UTF8.GetBytes(message), cancel);
+        }
+
+        /// <inheritdoc />
+        public Task PublishAsync(string topic, Stream message, CancellationToken cancel = default)
+        {
+            using (var ms = new MemoryStream())
+            {
+#pragma warning disable VSTHRD103
+                message.CopyTo(ms);
+#pragma warning disable VSTHRD103 
+                return PublishAsync(topic, ms.ToArray(), cancel);
+            }
+        }
+
+        /// <inheritdoc />
+        public Task PublishAsync(string topic, byte[] message, CancellationToken cancel = default)
+        {
+            var msg = CreateMessage(topic, message);
+            ++MesssagesPublished;
+            return Task
+               .WhenAll(Routers.Select(r => r.PublishAsync(msg, cancel)));
+        }
+
+        /// <inheritdoc />
+        public async Task SubscribeAsync(string topic,
+            Action<IPublishedMessage> handler,
+            CancellationToken cancellationToken)
+        {
+            var topicHandler = new TopicHandler {Topic = topic, Handler = handler};
+            _topicHandlers.TryAdd(topicHandler, topicHandler);
+
+            // TODO: need a better way.
+#pragma warning disable VSTHRD101 
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not suppress issues.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)